compiler/damlc/daml-preprocessor/src/DA/Daml/Preprocessor/TemplateConstraint.hs

```diff
@@ -0,0 +1,83 @@
+-- Copyright (c) 2019 The DAML Authors. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+-- | Every `template T a_1 a_n` gets desugared into a `class TInstance a_1 ... a_n`,
+-- an instance
+-- ```
+-- (*)    instance TInstance a_1 ... a_n => Template (T a_1 ... a_n)
+-- ```
+-- and instances `instance TInstance a_1 ... a_n => Choice (T a_1 ... a_n) C R`
+-- for each choice `C`.
+--
+-- Thus, a _generic_ exercise of a choice on `T t_1 ... t_n` needs to have the
+-- `TInstance t_1 ... t_n` constraint in scope. However, we want to keep the
+-- existence of the `TInstance` class an implementation detail and not expose
+-- it to our users. Instead we want our users to add a `Template (T t_1 ... t_n)`
+-- constraint when they want to perform the generic exercise.
+--
+-- Due to the (*) instance above, the constraint `Template (T t_1 ... t_n)`
+-- is satisfied if and only if the constraint `TInstance t_1 ... t_n` is
+-- satisfied. For the intent described above it would be necessary that GHC
+-- conlcudes the latter from the former. Unfortunately, GHC's type system only
+-- allows for concluding the former from the latter.
+--
+-- Thus, we add a preprocessing step which rewrites all constraints of
+-- the form `Template (T t_1 ... t_n)` into `TInstance t_1 ... t_n`.
+module DA.Daml.Preprocessor.TemplateConstraint (
+    templateConstraintPreprocessor
+    ) where
+
+import "ghc-lib" GHC
+import "ghc-lib-parser" BasicTypes
+import "ghc-lib-parser" OccName
+import "ghc-lib-parser" RdrName
+
+import Data.Generics.Uniplate.Data
+
+templateConstraintPreprocessor :: ParsedSource -> ParsedSource
+templateConstraintPreprocessor = fmap onModule
+
+onModule :: HsModule GhcPs -> HsModule GhcPs
+onModule = transformBi onType . transformBi onTyClDecl
+
+-- | The contexts of class and data definitions do not live inside an
+-- `HsType` but in their respective AST nodes. That's why we need to chase
+-- them down separately.
+onTyClDecl :: TyClDecl GhcPs -> TyClDecl GhcPs
+onTyClDecl decl = case decl of
+    ClassDecl{tcdCtxt} -> decl{tcdCtxt = fmap onContext tcdCtxt}
+    DataDecl{tcdDataDefn = defn@HsDataDefn{dd_ctxt}} -> decl{tcdDataDefn = defn{dd_ctxt = fmap onContext dd_ctxt}}
+    _ -> decl
+
+onType :: HsType GhcPs-> HsType GhcPs
+onType = \case
+    HsQualTy ext ctxt t -> HsQualTy ext (fmap onContext ctxt) (fmap onType t)
+    t -> descend onType t
+
+onContext :: HsContext GhcPs -> HsContext GhcPs
+onContext = map . fmap $ \case
+    HsAppTy _ (L _ (HsTyVar _ _ (L _ (occNameString . rdrNameOcc -> "Template")))) (L _ t)
```

Comment:
Can we at least check the module here as well?

Reply:
The module name would be the alias the user picks when importing the module containing the `Template` class. Thus, there's nothing we could really check for.